test/videoframeprocessor/DefaultVideoFrameProcessorPipeline.test.ts

```diff
@@ -0,0 +1,432 @@
+  let proc: VideoFrameProcessor;
+
+  class MockObserver implements VideoFrameProcessorPipelineObserver {
+    processingDidFailToStart = sinon.stub();
+    processingDidStop = sinon.stub();
+    processingLatencyTooHigh = sinon.stub();
+    processingDidStart = sinon.stub();
+  }
+
+  beforeEach(() => {
+    domMockBehavior = new DOMMockBehavior();
+    domMockBuilder = new DOMMockBuilder(domMockBehavior);
+    domMockBehavior.createElementCaptureStream = new MediaStream();
+    mockVideoStream = new MediaStream();
+    // @ts-ignore
+    mockVideoStream.id = mockStreamId;
+    // @ts-ignore
+    mockVideoTrack = new MediaStreamTrack('attach-media-input-task-video-track-id', 'video');
+    mockVideoStream.addTrack(mockVideoTrack);
+    proc = new NoOpVideoFrameProcessor();
+    pipe = new DefaultVideoFrameProcessorPipeline(logger, [proc]);
+  });
+
+  afterEach(() => {
+    if (domMockBuilder) {
+      domMockBuilder.cleanup();
+    }
+  });
+
+  describe('construction', () => {
+    it('can be constructed', () => {
+      assert.exists(pipe);
+    });
+  });
+
+  describe('setInputMediaStream', () => {
+    it('can set the input', async () => {
+      await pipe.setInputMediaStream(mockVideoStream);
+      const outputStream = await pipe.getInputMediaStream();
+      expect(outputStream.id).to.equal(mockStreamId);
+      await pipe.setInputMediaStream(null);
+    });
+
+    it('can only set MediaStream with video tracks', async () => {
+      const emptyStream = new MediaStream();
+      await pipe.setInputMediaStream(emptyStream);
+      const outputStream = await pipe.getInputMediaStream();
+      expect(outputStream).to.equal(null);
+      await pipe.setInputMediaStream(null);
+    });
+
+    it('can stop the pipeline multiple times', async () => {
+      await pipe.setInputMediaStream(null);
+      const outputStream = await pipe.getInputMediaStream();
+      expect(outputStream).to.equal(null);
+      await pipe.setInputMediaStream(null);
+    });
+
+    it('can start the pipeline with valid stream and stop with null', async () => {
+      const pipeObserver = new MockObserver();
+      const procs = [new NoOpVideoFrameProcessor()];
+      pipe.processors = procs;
+      pipe.addObserver(pipeObserver);
+      await pipe.setInputMediaStream(mockVideoStream);
+      await pipe.setInputMediaStream(null);
+    });
+
+    it('can start the pipeline with valid stream and stop with null', done => {
+      const pipeObserver = new MockObserver();
+      const procs = [new NoOpVideoFrameProcessor()];
+      pipe.processors = procs;
+      pipe.addObserver(pipeObserver);
+      pipe.setInputMediaStream(mockVideoStream);
+      new TimeoutScheduler(300).start(() => {
+        expect(pipeObserver.processingDidStart.called).to.eq(true);
+        pipe.setInputMediaStream(null);
+      });
+
+      new TimeoutScheduler(350).start(() => {
+        expect(pipeObserver.processingDidStop.called).to.eq(true);
+        done();
+      });
```

Comment:
Are you doing this because you're trying to wait for the observers to have been called?

If so, scheduling like this is absolutely not the way to do it, unless you're trying to test that 100% of the time this callback will have been invoked within 350ms.

Either use promises or callbacks to have `done` called or have the test return when both observer methods are called. (See `Promise.all`.)